### PR TITLE
Massively improve Stalker's reliability on WoW64

### PR DESCRIPTION
**Pull request by zuypt**
My windows version is 1909 (OS Build 18363.720). Beside `ntdll!Wow64Transition` there is also `kernel32!Wow64Transition` but both of them point to the same heaven gate address. `kernel32!Wow64SystemServiceCall` and `ntdll!Wow64SystemServiceCall` each jump to their own `Wow64Transition`. I'm not sure if older windows versions have `kernel32!Wow64Transition` and which one gonna be used when. [](https://github.com/frida/frida-gum/pull/333) only checks for jump to ntdll!Wow64Transition hence kernel32!Wow64Transition case is missed. `LoadLibrary` uses kernel!Wow64SystemServiceCall therefore stalker breaks everytime `LoadLibrary` is  called.

### CODE REVIEW

File: gum/backend-x86/gumstalker-x86.c [hunk 1]
```diff
@@ -2801,9 +2805,8 @@ gum_exec_block_virtualize_branch_insn (GumExecBlock * block,
   else if (op->type == X86_OP_MEM)
   {
 #if GLIB_SIZEOF_VOID_P == 4 && defined (HAVE_WINDOWS)
-    if (ctx->stalker->wow64_transition_address != NULL &&
-        op->mem.disp == GPOINTER_TO_UINT (
-            ctx->stalker->wow64_transition_address) &&
+    if (ctx->stalker->ntdll_wow64_transition_address != NULL &&
+        (op->mem.disp == GPOINTER_TO_UINT (ctx->stalker->ntdll_wow64_transition_address) || (ctx->stalker->kernel32_wow64_transition_address != NULL && op->mem.disp == GPOINTER_TO_UINT (ctx->stalker->kernel32_wow64_transition_address))) &&
```
Comment (oleavr) [inline gum/backend-x86/gumstalker-x86.c]: Exceeds 80 cols. Style is documented [here](https://github.com/frida/frida/blob/master/CONTRIBUTING.md).

Comment (zuypt) [reply gum/backend-x86/gumstalker-x86.c]: Fixed, also I think this is the more correct way

File: gum/backend-x86/gumstalker-x86.c [hunk 2]
```diff
@@ -3058,7 +3068,7 @@ gum_exec_block_virtualize_sysenter_insn (GumExecBlock * block,
 
 static GumVirtualizationRequirements
 gum_exec_block_virtualize_wow64_transition (GumExecBlock * block,
-                                            GumGeneratorContext * gc)
+                                            GumGeneratorContext * gc, gpointer wow64_transition_addr)
```
Comment (oleavr) [inline gum/backend-x86/gumstalker-x86.c]: The added argument should go on the next line as per the style guide.

File: gum/backend-x86/gumstalker-x86.c [hunk 3]
```diff
@@ -2801,14 +2805,20 @@ gum_exec_block_virtualize_branch_insn (GumExecBlock * block,
...
+      if (ctx->stalker->ntdll_wow64_transition_address != NULL &&
+          op->mem.disp == GPOINTER_TO_UINT (ctx->stalker->ntdll_wow64_transition_address)) 
+      {
+        return gum_exec_block_virtualize_wow64_transition (block, gc, ctx->stalker->ntdll_wow64_transition_address);
+      }
+      else if (ctx->stalker->kernel32_wow64_transition_address != NULL &&
+          op->mem.disp == GPOINTER_TO_UINT (ctx->stalker->kernel32_wow64_transition_address))
+      {
+        return gum_exec_block_virtualize_wow64_transition (block, gc, ctx->stalker->kernel32_wow64_transition_address);
+      }
```
Comment (oleavr) [inline gum/backend-x86/gumstalker-x86.c]: Here and above it looks like multiple lines exceed 80 cols.

Comment (zuypt) [reply gum/backend-x86/gumstalker-x86.c]: i think i got it right this time, hopefully

File: gum/backend-x86/gumstalker-x86.c [hunk 4]
```diff
@@ -2801,14 +2805,24 @@ gum_exec_block_virtualize_branch_insn (GumExecBlock * block,
...
-        op->mem.disp == GPOINTER_TO_UINT (
-            ctx->stalker->wow64_transition_address) &&
-        op->mem.segment == X86_REG_INVALID &&
+    if (op->mem.segment == X86_REG_INVALID &&
         op->mem.base == X86_REG_INVALID &&
         op->mem.index == X86_REG_INVALID)
     {
-      return gum_exec_block_virtualize_wow64_transition (block, gc);
+      if (ctx->stalker->ntdll_wow64_transition_address != NULL &&
+          op->mem.disp == GPOINTER_TO_UINT (
```
Comment (oleavr) [inline gum/backend-x86/gumstalker-x86.c]:
```suggestion
          op->mem.disp == GPOINTER_TO_SIZE (
```

File: gum/backend-x86/gumstalker-x86.c [hunk 5]
```diff
@@ -2801,14 +2805,24 @@ gum_exec_block_virtualize_branch_insn (GumExecBlock * block,
...
+            ctx->stalker->ntdll_wow64_transition_address)) 
+      {
+        return gum_exec_block_virtualize_wow64_transition (block, gc,
+          ctx->stalker->ntdll_wow64_transition_address);
+      }
+      else if (ctx->stalker->kernel32_wow64_transition_address != NULL &&
+          op->mem.disp == GPOINTER_TO_UINT (
```
Comment (oleavr) [inline gum/backend-x86/gumstalker-x86.c]:
```suggestion
          op->mem.disp == GPOINTER_TO_SIZE (
```

File: gum/backend-x86/gumstalker-x86.c [hunk 6]
```diff
@@ -2801,14 +2805,24 @@ gum_exec_block_virtualize_branch_insn (GumExecBlock * block,
...
-            ctx->stalker->wow64_transition_address) &&
-        op->mem.segment == X86_REG_INVALID &&
+    if (op->mem.segment == X86_REG_INVALID &&
         op->mem.base == X86_REG_INVALID &&
         op->mem.index == X86_REG_INVALID)
     {
-      return gum_exec_block_virtualize_wow64_transition (block, gc);
+      if (ctx->stalker->ntdll_wow64_transition_address != NULL &&
+          op->mem.disp == GPOINTER_TO_UINT (
+            ctx->stalker->ntdll_wow64_transition_address))
```
Comment (oleavr) [inline gum/backend-x86/gumstalker-x86.c]:
```suggestion
              ctx->stalker->ntdll_wow64_transition_address)) 
```

File: gum/backend-x86/gumstalker-x86.c [hunk 7]
```diff
@@ -2801,14 +2805,24 @@ gum_exec_block_virtualize_branch_insn (GumExecBlock * block,
...
+      if (ctx->stalker->ntdll_wow64_transition_address != NULL &&
+          op->mem.disp == GPOINTER_TO_UINT (
+            ctx->stalker->ntdll_wow64_transition_address)) 
+      {
+        return gum_exec_block_virtualize_wow64_transition (block, gc,
+          ctx->stalker->ntdll_wow64_transition_address);
+      }
+      else if (ctx->stalker->kernel32_wow64_transition_address != NULL &&
+          op->mem.disp == GPOINTER_TO_UINT (
+            ctx->stalker->kernel32_wow64_transition_address))
```
Comment (oleavr) [inline gum/backend-x86/gumstalker-x86.c]:
```suggestion
              ctx->stalker->kernel32_wow64_transition_address))
```

File: gum/backend-x86/gumstalker-x86.c [hunk 8]
```diff
@@ -2801,14 +2805,24 @@ gum_exec_block_virtualize_branch_insn (GumExecBlock * block,
...
         op->mem.base == X86_REG_INVALID &&
         op->mem.index == X86_REG_INVALID)
     {
-      return gum_exec_block_virtualize_wow64_transition (block, gc);
+      if (ctx->stalker->ntdll_wow64_transition_address != NULL &&
+          op->mem.disp == GPOINTER_TO_UINT (
+            ctx->stalker->ntdll_wow64_transition_address)) 
+      {
+        return gum_exec_block_virtualize_wow64_transition (block, gc,
+          ctx->stalker->ntdll_wow64_transition_address);
```
Comment (oleavr) [inline gum/backend-x86/gumstalker-x86.c]:
```suggestion
            ctx->stalker->ntdll_wow64_transition_address);
```

File: gum/backend-x86/gumstalker-x86.c [hunk 9]
```diff
@@ -2801,14 +2805,24 @@ gum_exec_block_virtualize_branch_insn (GumExecBlock * block,
...
+      }
+      else if (ctx->stalker->kernel32_wow64_transition_address != NULL &&
+          op->mem.disp == GPOINTER_TO_UINT (
+            ctx->stalker->kernel32_wow64_transition_address))
+      {
+        return gum_exec_block_virtualize_wow64_transition (block, gc,
+          ctx->stalker->kernel32_wow64_transition_address);
```
Comment (oleavr) [inline gum/backend-x86/gumstalker-x86.c]:
```suggestion
            ctx->stalker->kernel32_wow64_transition_address);
```

**Comment by zuypt** [comment]
The issue is not done here. I just discovered `win32u!Wow64SystemServiceCall` that jump to `win32u!Wow64Transition`. Looks like each DLL has its own `Wow64Transition`. Need to find a more generic way fix this issue. Please don't merge for now

**Comment by zuypt** [comment]
Using windbg to attach to a 32bit notepad process.
```
0:007> x *!Wow64Transition
75c44fc0          USER32!Wow64Transition = <no type information>
75c44fc0          USER32!Wow64Transition = <no type information>
76129adc          KERNELBASE!Wow64Transition = <no type information>
76129adc          KERNELBASE!Wow64Transition = <no type information>
761f4010          win32u!Wow64Transition = <no type information>
761f4010          win32u!Wow64Transition = <no type information>
778b1f90          KERNEL32!Wow64Transition = <no type information>
778b1f90          KERNEL32!Wow64Transition = <no type information>
77d91228          ntdll!Wow64Transition = <no type information>
```
Should I handle all these cases ?

**Comment by oleavr** [comment]
@zuypt Ooh nice find. That would be great! 🤘

**Comment by zuypt** [comment]
> @zuypt Ooh nice find. That would be great! 🤘

i pushed it can you check.

File: gum/backend-x86/gumstalker-x86.c [hunk 10]
```diff
@@ -2801,14 +2817,46 @@ gum_exec_block_virtualize_branch_insn (GumExecBlock * block,
...
+        return gum_exec_block_virtualize_wow64_transition (block, gc,
+            ctx->stalker->user32_wow64_transition_address);
+      }
+      else if (ctx->stalker->win32u_wow64_transition_address != NULL &&
+          op->mem.disp == GPOINTER_TO_SIZE (
+              ctx->stalker->win32u_wow64_transition_address))
+      {
+        return gum_exec_block_virtualize_wow64_transition (block, gc,
+            ctx->stalker->win32u_wow64_transition_address);
+      }
```
Comment (oleavr) [inline gum/backend-x86/gumstalker-x86.c:2858]: Perhaps it would be clearer to use a loop here.

File: gum/backend-x86/gumstalker-x86.c [hunk 11]
```diff
@@ -2801,14 +2817,46 @@ gum_exec_block_virtualize_branch_insn (GumExecBlock * block,
...
+
```
Comment (oleavr) [inline gum/backend-x86/gumstalker-x86.c:2859]: Left-over blank line.

File: gum/backend-x86/gumstalker-x86.c (hunk 10 above)
Comment (zuypt) [reply gum/backend-x86/gumstalker-x86.c:2858]: Please spare me this loop. Will definitely add a loop if there is more than 8 case

**Comment by oleavr** [comment]
Thanks! Don't worry, I'll do the final tweaks before merging. Currently heads down updating dependencies, will get to this before the next release.

**Comment by oleavr** [comment]
(Still working on the dependencies, but getting there.)

**Comment by oleavr** [comment]
Landed with some minor tweaks in b8d6b2a. Thanks! 🤘

**Pull request closed without merging**